mtags/src/main/scala-2/scala/meta/internal/pc/InferredMethodProvider.scala

```diff
@@ -0,0 +1,496 @@
+      val listSymbol = context.lookupSymbol(argumentList, _ => true)
+      if (listSymbol.isSuccess) {
+        listSymbol.symbol.tpe match {
+          // need to find the type of the value on which we are mapping
+          case TypeRef(_, _, TypeRef(_, inputType, _) :: Nil) =>
+            val paramsString = s"arg0: ${prettyType(inputType.tpe)}"
+            signature(
+              name = errorMethod.name.toString(),
+              paramsString,
+              None,
+              identity,
+              None
+            )
+          case NullaryMethodType(
+                TypeRef(_, _, TypeRef(_, inputType, _) :: Nil)
+              ) =>
+            val paramsString = s"arg0: ${prettyType(inputType.tpe)}"
+            signature(
+              name = errorMethod.name.toString(),
+              paramsString,
+              None,
+              identity,
+              None
+            )
+          case _ =>
+            Nil
+        }
+      } else Nil
+    }
+    def makeEditsMethodObject(
+        arguments: List[Tree],
+        container: Name,
+        errorMethod: Select
+    ): List[(FileURI, TextEdit)] = {
+      def makeClassMethodEdits(
+          template: Template
+      ): List[(FileURI, TextEdit)] = {
+        val insertPos: Position =
+          inferEditPosition(template)
+        val templateFileUri = template.pos.source.content
+          .map(_.toString)
+          .mkString
+        if (params.uri().toString() != getPositionUri(insertPos)) {
+          println("You can only infer method for current file.")
+          Nil
+        } else
+          signature(
+            name = errorMethod.name.toString(),
+            paramsString = argumentsString(arguments).getOrElse(""),
+            retType = None,
+            postProcess = method => {
+              if (
+                hasBody(
+                  templateFileUri,
+                  template
+                ).isDefined
+              )
+                s"\n  $method"
+              else s" {\n  $method}"
+            },
+            position = Option(insertPos)
+          )
+      }
+
+      // we need to get the type of the container of our undefined method
+      val containerSymbol = context.lookupSymbol(container, _ => true)
+
+      if (containerSymbol.isSuccess) {
+        (if (containerSymbol.symbol.tpe.isInstanceOf[NullaryMethodType])
+           containerSymbol.symbol.tpe
+             .asInstanceOf[NullaryMethodType]
+             .resultType
+         else containerSymbol.symbol.tpe) match {
+          case TypeRef(_, classSymbol, _) =>
+            // we get the position of the container
+            // class because we want to add the method
+            // definition there
+            val containerClass =
+              context.lookupSymbol(classSymbol.name, _ => true)
+            // this gives me the position of the class
+            typedTreeAt(containerClass.symbol.pos) match {
+              // class case
+              case ClassDef(
```

Comment:
Could we use ImplDef superclass for both ClassDef and ModuleDef?

Reply:
it seems a good idea but a simple ClassDef -> ImplDef didn't compile.
I will look into the other issues you have highlighted since this feels minor and requires digging time.